### DISCUSSION TITLE
debugging

### DISCUSSION THREAD

**Discussion by istvan-b01** [spakin/SimpInkScr#93] [General]
Due to the nature of the tools I hardly imagine but is there a facility to debug the code as there are no line numbers provided if something goes wrong which makes is pretty painful to figure a bug out. At some case I had do literally blanket every line with some print statement to try to find out where it breaks.
Just asking in case I'm not aware, I imagine that there might be an import that would allow a code to be run in an IDE maybe...

**Comment by spakin** [comment]
To be honest, I normally use `print` for debugging.

A contributor to Simple Inkscape Scripting supplied pull request #55, which supposedly provides support for the Visual Studio Code IDE, but I've never tested that myself and therefore don't know if I've since made some code modification that has broken VSCode integration.

It's possible to run Simple Inkscape Scripting from the command line.  Set your `PYTHONPATH` to include the directory containing Inkscape's `inkex` Python module and run `simple_inkscape_scripting.py`:
```bash
env PYTHONPATH=/usr/share/inkscape/extensions:$PYTHONPATH python ~/.config/inkscape/extensions/simple_inkscape/scripting/simpinkscr/simple_inkscape_scripting.py --py-source=my-script.py /usr/share/inkscape/templates/default.svg
```
If nothing else, that will let you use Python's [pdb](https://docs.python.org/3/library/pdb.html) as a debugger:
```bash
env PYTHONPATH=/usr/share/inkscape/extensions:$PYTHONPATH python -m pdb ~/.config/inkscape/extensions/simple_inkscape/scripting/simpinkscr/simple_inkscape_scripting.py --py-source=my-script.py /usr/share/inkscape/templates/default.svg
```

**Reply by istvan-b01** [reply]
Thanks for the quick reply, I certainly will give a go but the problem is that unfortunately (me using mint) the default choice for installing inkscape is v 1.1.2 and that breaks somehow the code (broke on 1.1.2 but run on windows version 1.2) so today I installed the 1.3 flapack one which, as far as I understood, is looking for the extensions at different places (snap store installed it-1.2.2, but then couldn't see it, so it didn't show up in the render part. in the software packager (1.1.2) and in flatpack it does show up twice).
anyway, I'll look into it.
the feedback does provide some location but somehow it calculates the line number in a not very useful way as it seem to have an offset. I don't know how it calculates them. does it count empty lines? 2 statements in 1 line counts as 1 or 2? etc...

slightly changing topic  ad now I'm being a bit greedy and have more questions
- when ever I use 'apply_path_operation', both windows and linux version initiate a new window opening which then closes down. is straight after. it this normal or avoidable. It does not bother me bur if I generate a few hundred layers, each having 2-3 operation like this then it does slow down the process significantly.
- even less important, how can I change the label of an object in a script?
- most importantly (related to the above question) I've seen some questions that are far more advanced than anything I asked, is there a documentation of the object model (of Inkscape I guess) so I could look things up myself before asking? and is there documentation listing let's say all the methods of SIS (if I may abbreviate) so I could extend my tool set as at the moment I feel like I'm using Java but with the tool set that BrainF@ck offers.

**Reply by spakin** [reply]
I don't know why the line numbers might be counted oddly.  I haven't used the debugger much myself on Simple Inkscape Scripting scripts.

With respect to your new questions,
* The new window is normal.  [`apply_path_operation`](https://github.com/spakin/SimpInkScr/wiki/Path-operations#applying-path-operations) needs to spawn a separate Inkscape process to perform the path operations as these aren't supported directly in Python.
* Use [`svg_set`](https://github.com/spakin/SimpInkScr/wiki/Modifying-existing-objects#svg-level-attribute-modifications) to set the label:
```Python
c = circle((100, 100), 75, fill='cyan')
c.svg_set('inkscape:label', 'My circle')
```
* The [Simple Inkscape Scripting wiki](https://github.com/spakin/SimpInkScr/wiki) documents all functions, methods, variables, and properties defined by Simple Inkscape Scripting.  Simple Inkscape Scripting's objects wrap those provided by Inkscape's [`inkex`](https://inkscape.gitlab.io/extensions/documentation/) library to provide more convenient access.

**Reply by istvan-b01** [reply]
very much appreciated!!!
Your answers are spot on, giving full package, context and great examples!
Thanks.
